README
MemPool library 1.0
===================


The MemPool library is a stand-alone memory allocation library.
It is a block allocator library that uses a single fixed size memory
pool to allocate variable sized memory blocks, and exports functions
similar to malloc, realloc and free.
It is intended to either simulate low memory targets on GNU/Linux and
other systems, or to completely replace the system memory management.

It is based on the memory management code from the eCos operating
system, which was cut free from the environment so it can freely be
used on GNU/Linux and other systems.

It is re-published under the eCos licence.


How to build
============

GNU/Linux:

cd ./build
../MemPool/configure
make && make test

Windows:

cd .\build
cmake ..\MemPool
cmake --build .


How to use
==========

The include file "mempool.h" contains all exported functions.
All functions in this library are thread-safe.


cyg_bool MemPool_init(cyg_uint32)

- takes the requested size in bytes of the memory pool (up to 2GB)
  and returns true on success false if not enough memory.
- call this function always first.


void MemPool_exit(void)

- shuts the MemPool library down.
- all previously allocated memory gets released to the O/S.
- call this function optionally on shutdown.


void MemPool_get_status(cyg_uint32 flags, struct Cyg_Mempool_Status* status)

- returns current status about MemPool.
- flags is CYG_MEMPOOL_STAT_ALL or a combination of CYG_MEMPOOL_STAT_XXX (see mpcommon.h).
- *status is filled on return, and contains different usage statistics.


void* MemPool_Memory_Alloc_Func(cyg_uint32 size)

- allocates a variable sized memory block out of the MemPool.
- can be used to replace the malloc function.


void* MemPool_Memory_ReAlloc_Func(void* mem, cyg_uint32 new_size)

- reallocates a variable sized memory block out of the MemPool.
- can be used to replace the realloc function.


void  MemPool_Memory_Free_Func(void* mem)

- releases a variable sized memory block.
- can be used to replace the free function.


Copying
=======

Copying and distribution of this file, with or without modification,
are permitted in any medium without royalty provided the copyright
notice and this notice are preserved.  This file is offered as-is,
without any warranty.


NEWS
News
====

- On 2017-06-20 MemPool was dubbed as GNU package by RMS.


Copying and distribution of this file, with or without modification,
are permitted in any medium without royalty provided the copyright
notice and this notice are preserved.  This file is offered as-is,
without any warranty.


common.h
// This file is part of the MemPool library.
// Copyright (C) 2017 Free Software Foundation, Inc.
//
// MemPool is free software; you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free
// Software Foundation; either version 2 or (at your option) any later
// version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software Foundation, Inc.,
// 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
//
// As a special exception, if other files instantiate templates or use
// macros or inline functions from this file, or you compile this file
// and link it with other works to produce a work based on this file,
// this file does not by itself cause the resulting work to be covered by
// the GNU General Public License.  However the source code for this file
// must still be made available in accordance with section (3) of the GNU
// General Public License v2.
//
// This exception does not invalidate any other reasons why a work based
// on this file might be covered by the GNU General Public License.
#ifndef CYGONCE_MEMALLOC_COMMON_H
#define CYGONCE_MEMALLOC_COMMON_H

/*==========================================================================
//
//      common.h
//
//      Shared definitions used by memory allocators
//
//==========================================================================
//#####DESCRIPTIONBEGIN####
//
// Author(s):    jlarmour
// Contributors: Bernd Edlinger
// Date:         2000-06-12
// Purpose:      Shared definitions used by memory allocators
// Description:
// Usage:        #include <cyg/memalloc/common.hxx>
//
//
//####DESCRIPTIONEND####
//
//========================================================================*/

#if defined _WIN32 || defined _WIN32_WCE
#define MEMPOOL_WIN
#else
#define MEMPOOL_POSIX
#endif

#ifdef MEMPOOL_WIN
#define MEMPOOL_API_CALL __stdcall
#else
#define MEMPOOL_API_CALL
#endif

/* TYPE DEFINITIONS */

typedef unsigned char      cyg_uint8;
typedef unsigned int       cyg_uint32;
typedef int       cyg_bool;

/*==========================================================================
// struct Cyg_Mempool_Status is returned by the get_status() method of
// standard eCos memory allocators. After return from get_status(), any
// field of type T may be set to ((T)-1) to indicate that the information
// is not available or not applicable to this allocator.
//========================================================================*/


struct Cyg_Mempool_Status {
    const cyg_uint8 *arenabase; /* base address of entire pool */
    cyg_uint32  arenasize;      /* total size of entire pool */
    cyg_uint32  freeblocks;     /* number of chunks free for use */
    cyg_uint32  totalallocated; /* total allocated space in bytes */
    cyg_uint32  totalfree;      /* total space in bytes not in use */
    cyg_uint32  blocksize;      /* block size if fixed block */
    cyg_uint32  maxfree;        /* size of largest unused block */
    cyg_uint8   waiting;        /* are there any threads waiting for memory? */
    const cyg_uint8 *origbase;  /* address of original region used when pool */
                                /* created */
    cyg_uint32  origsize;       /* size of original region used when pool */
                                /* created */

    /*==========================================================================
    // maxoverhead is the *maximum* per-allocation overhead imposed by
    // the allocator implementation. Note: this is rarely the typical
    // overhead which often depends on the size of the allocation requested.
    // It includes overhead due to alignment constraints. For example, if
    // maxfree and maxoverhead are available for this allocator, then an
    // allocation request of (maxfree-maxoverhead) bytes must always succeed
    // Unless maxoverhead is set to -1 of course, in which case the allocator
    // does not support reporting this information.
    //========================================================================*/

    cyg_uint8   maxoverhead;
    cyg_uint32  maxAllocBytes;
    cyg_uint32  actualAllocBytes;

};

/*==========================================================================
// Flags to pass to get_status() methods to tell it which stat(s) is/are
// being requested
//========================================================================*/

#define CYG_MEMPOOL_STAT_ARENABASE       (1<<0)
#define CYG_MEMPOOL_STAT_ARENASIZE       (1<<1)
#define CYG_MEMPOOL_STAT_FREEBLOCKS      (1<<2)
#define CYG_MEMPOOL_STAT_TOTALALLOCATED  (1<<3)
#define CYG_MEMPOOL_STAT_TOTALFREE       (1<<4)
#define CYG_MEMPOOL_STAT_BLOCKSIZE       (1<<5)
#define CYG_MEMPOOL_STAT_MAXFREE         (1<<6)
#define CYG_MEMPOOL_STAT_WAITING         (1<<7)
#define CYG_MEMPOOL_STAT_ORIGBASE        (1<<9)
#define CYG_MEMPOOL_STAT_ORIGSIZE        (1<<10)
#define CYG_MEMPOOL_STAT_MAXOVERHEAD     (1<<11)

#define CYG_MEMPOOL_STAT_ALL                    \
              ( CYG_MEMPOOL_STAT_ARENABASE      \
              | CYG_MEMPOOL_STAT_ARENASIZE      \
              | CYG_MEMPOOL_STAT_FREEBLOCKS     \
              | CYG_MEMPOOL_STAT_TOTALALLOCATED \
              | CYG_MEMPOOL_STAT_TOTALFREE      \
              | CYG_MEMPOOL_STAT_BLOCKSIZE      \
              | CYG_MEMPOOL_STAT_MAXFREE        \
              | CYG_MEMPOOL_STAT_WAITING        \
              | CYG_MEMPOOL_STAT_ORIGBASE       \
              | CYG_MEMPOOL_STAT_ORIGSIZE       \
              | CYG_MEMPOOL_STAT_MAXOVERHEAD )

#endif /* ifndef CYGONCE_MEMALLOC_COMMON_H */
/* EOF common.h */


mempool.c
/*==========================================================================
//
// This file is part of the MemPool library.
// Copyright (C) 2017 Bernd Edlinger
//
// MemPool is free software; you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free
// Software Foundation; either version 3 or (at your option) any later
// version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software Foundation, Inc.,
// 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
//
// As a special exception, if other files instantiate templates or use
// macros or inline functions from this file, or you compile this file
// and link it with other works to produce a work based on this file,
// this file does not by itself cause the resulting work to be covered by
// the GNU General Public License.  However the source code for this file
// must still be made available in accordance with section (6) of the GNU
// General Public License v3.
//
// This exception does not invalidate any other reasons why a work based
// on this file might be covered by the GNU General Public License.
//
//========================================================================*/

#include "mpcommon.h"

#ifdef MEMPOOL_WIN
#include <windows.h>
#endif

#ifdef MEMPOOL_POSIX
/* The most common include on Linux */
#include <errno.h>
#include <ctype.h>
#include <fcntl.h>
#include <unistd.h>
#include <pthread.h>
#include <time.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mman.h>
#endif

#include "dlmalloc.h"
#include "mempool.h"

#ifdef MEMPOOL_POSIX
#define CRITICAL_SECTION pthread_mutex_t
#define INITIALIZE_CRITICAL_SECTION(pcs) pthread_mutex_init(pcs, NULL)
#define DELETE_CRITICAL_SECTION(pcs) pthread_mutex_destroy(pcs)
#define ENTER_CRITICAL_SECTION(pcs) pthread_mutex_lock(pcs)
#define LEAVE_CRITICAL_SECTION(pcs) pthread_mutex_unlock(pcs)
#endif

#ifdef MEMPOOL_WIN
#define INITIALIZE_CRITICAL_SECTION(pcs) InitializeCriticalSection(pcs)
#define DELETE_CRITICAL_SECTION(pcs) DeleteCriticalSection(pcs)
#define ENTER_CRITICAL_SECTION(pcs) EnterCriticalSection(pcs)
#define LEAVE_CRITICAL_SECTION(pcs) LeaveCriticalSection(pcs)
#endif

static CRITICAL_SECTION cs_mem;

cyg_bool
MEMPOOL_API_CALL MemPool_init(cyg_uint32 size)
{
  void* mem_block;
#ifdef MEMPOOL_WIN
  mem_block = VirtualAlloc(NULL, size, MEM_COMMIT|MEM_RESERVE, PAGE_READWRITE);
  if (mem_block == NULL)
  {
    return 0;
  }
#endif
#ifdef MEMPOOL_POSIX
  mem_block = mmap(NULL, size, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0);
  if (mem_block == MAP_FAILED)
  {
    return 0;
  }
#endif
  Cyg_dlmalloc_init((cyg_uint8*)mem_block, size);
  INITIALIZE_CRITICAL_SECTION(&cs_mem);
  return 1;
}

void
MEMPOOL_API_CALL MemPool_exit(void)
{
  struct Cyg_Mempool_Status status;
  Cyg_dlmalloc_get_status(0, &status);
#ifdef MEMPOOL_WIN
  VirtualFree((void*)status.origbase, 0, MEM_RELEASE);
#endif
#ifdef MEMPOOL_POSIX
  munmap((void*)status.origbase, status.origsize);
#endif
  DELETE_CRITICAL_SECTION(&cs_mem);
}

void
MEMPOOL_API_CALL MemPool_get_status(cyg_uint32 flags, struct Cyg_Mempool_Status* status)
{
  memset(status, -1, sizeof(*status));
  ENTER_CRITICAL_SECTION(&cs_mem);
  Cyg_dlmalloc_get_status(flags, status);
  LEAVE_CRITICAL_SECTION(&cs_mem);
}

void*
MEMPOOL_API_CALL MemPool_Memory_Alloc_Func(cyg_uint32 size)
{
  void* new_mem;
  ENTER_CRITICAL_SECTION(&cs_mem);
  new_mem = Cyg_dlmalloc_try_alloc(size);
  LEAVE_CRITICAL_SECTION(&cs_mem);
  return new_mem;
}

void*
MEMPOOL_API_CALL MemPool_Memory_ReAlloc_Func(void* mem, cyg_uint32 new_size)
{
  if (mem == NULL)
  {
    return MemPool_Memory_Alloc_Func(new_size);
  }
  else if (new_size == 0)
  {
    MemPool_Memory_Free_Func(mem);
    return NULL;
  }
  else
  {
    void* new_mem;
    cyg_uint32 old_size;
    ENTER_CRITICAL_SECTION(&cs_mem);
    new_mem = Cyg_dlmalloc_resize_alloc((cyg_uint8*)mem, new_size, &old_size);
    LEAVE_CRITICAL_SECTION(&cs_mem);
    if (new_mem == NULL)
    {
      new_mem = MemPool_Memory_Alloc_Func(new_size);
      if (new_mem != NULL)
      {
        memcpy(new_mem, mem, old_size);
        MemPool_Memory_Free_Func(mem);
      }
    }
    return new_mem;
  }
}

void
MEMPOOL_API_CALL MemPool_Memory_Free_Func(void* mem)
{
  if (mem != NULL)
  {
    ENTER_CRITICAL_SECTION(&cs_mem);
    Cyg_dlmalloc_free((cyg_uint8*)mem);
    LEAVE_CRITICAL_SECTION(&cs_mem);
  }
}


mempool.h
/*==========================================================================
//
// This file is part of the MemPool library.
// Copyright (C) 2017 Bernd Edlinger
//
// MemPool is free software; you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free
// Software Foundation; either version 3 or (at your option) any later
// version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software Foundation, Inc.,
// 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
//
// As a special exception, if other files instantiate templates or use
// macros or inline functions from this file, or you compile this file
// and link it with other works to produce a work based on this file,
// this file does not by itself cause the resulting work to be covered by
// the GNU General Public License.  However the source code for this file
// must still be made available in accordance with section (6) of the GNU
// General Public License v3.
//
// This exception does not invalidate any other reasons why a work based
// on this file might be covered by the GNU General Public License.
//
//========================================================================*/
#ifndef MEMPOOL_MEMPOOL_H
#define MEMPOOL_MEMPOOL_H

#include "mpcommon.h"

#ifdef __cplusplus
extern "C"
{
#endif
cyg_bool  MEMPOOL_API_CALL MemPool_init(cyg_uint32);
void  MEMPOOL_API_CALL MemPool_exit(void);
void  MEMPOOL_API_CALL MemPool_get_status(cyg_uint32, struct Cyg_Mempool_Status*);
void* MEMPOOL_API_CALL MemPool_Memory_Alloc_Func(cyg_uint32);
void* MEMPOOL_API_CALL MemPool_Memory_ReAlloc_Func(void*, cyg_uint32);
void  MEMPOOL_API_CALL MemPool_Memory_Free_Func(void*);

#ifdef __cplusplus
}
#endif

#endif /* ifndef MEMPOOL_MEMPOOL_H */
/* EOF mempool.h */


testapp.c
/*==========================================================================
//
// This file is part of the MemPool library.
// Copyright (C) 2017 Bernd Edlinger
//
// MemPool is free software; you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free
// Software Foundation; either version 3 or (at your option) any later
// version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software Foundation, Inc.,
// 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
//
// As a special exception, if other files instantiate templates or use
// macros or inline functions from this file, or you compile this file
// and link it with other works to produce a work based on this file,
// this file does not by itself cause the resulting work to be covered by
// the GNU General Public License.  However the source code for this file
// must still be made available in accordance with section (6) of the GNU
// General Public License v3.
//
// This exception does not invalidate any other reasons why a work based
// on this file might be covered by the GNU General Public License.
//
//========================================================================*/
#include "mempool.h"
#include "stddef.h"
#include "stdio.h"

void *my_malloc(size_t size)
{
    return MemPool_Memory_Alloc_Func(size);
}

void *my_realloc(void* data, size_t size)
{
    return MemPool_Memory_ReAlloc_Func(data, size);
}

void my_free(void* data)
{
    MemPool_Memory_Free_Func(data);
}

/* create fragmentaion */
void memtest(unsigned k)
{
    unsigned i;
    void **p,**q;
    p=q=my_malloc(504);
    for (i=0; i<k; i++)
    {
        q=*q=my_malloc(504+4*(i%16));
    }
    *q=0;
    q=p;
    for (i=0; i<k; i++)
    {
        void **n=*q;
        q=*q=*n;
        my_free(n);
        if (!q || !*q) q=p;
    }
    my_free(p);
}

#define MAX_HEAP_SIZE 0x100000
int main()
{
   struct Cyg_Mempool_Status MallocStatus;
   void *p;
   if(!MemPool_init(MAX_HEAP_SIZE))
   {
     printf("MemPool init failure\n");
     return 2;
   }
   p = my_malloc(1);
   my_free(p);
   p = my_malloc(MAX_HEAP_SIZE); /* will fail */
   p = my_malloc(100);
   p = my_realloc(p, 1000);
   my_free(p);
   memtest(1500);
   MemPool_get_status(0, &MallocStatus);
   printf("Allocated memory: %u (max) %u (act)\n", MallocStatus.maxAllocBytes, MallocStatus.actualAllocBytes);

   MemPool_exit();
   if (MallocStatus.actualAllocBytes != 0)
   {
     return 1;
   }
   return 0;
}


dlmalloc.h
/*==========================================================================
//
// This file is part of the MemPool library.
// Copyright (C) 2017 Bernd Edlinger
//
// MemPool is free software; you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free
// Software Foundation; either version 3 or (at your option) any later
// version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software Foundation, Inc.,
// 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
//
// As a special exception, if other files instantiate templates or use
// macros or inline functions from this file, or you compile this file
// and link it with other works to produce a work based on this file,
// this file does not by itself cause the resulting work to be covered by
// the GNU General Public License.  However the source code for this file
// must still be made available in accordance with section (6) of the GNU
// General Public License v3.
//
// This exception does not invalidate any other reasons why a work based
// on this file might be covered by the GNU General Public License.
//
//==========================================================================*/
#ifndef MEMPOOL_DLMALLOC_H
#define MEMPOOL_DLMALLOC_H

/*==========================================================================
//
//      dlmalloc.h
//
//      Interface to the port of Doug Lea's malloc implementation
//
//==========================================================================
//#####DESCRIPTIONBEGIN####
//
// Author(s):    jlarmour
// Contributors: Bernd Edlinger
// Date:         2000-06-18
// Purpose:      Define standard interface to Doug Lea's malloc implementation
// Description:  Doug Lea's malloc has been ported to eCos. This file provides
//               the interface between the implementation and the standard
//               memory allocator interface required by eCos
// Usage:        #include <cyg/memalloc/dlmalloc.hxx>
//
//
//####DESCRIPTIONEND####
//
//==========================================================================*/

/* CONFIGURATION */

/* INCLUDES */

#include "mpcommon.h"

#ifdef __cplusplus
extern "C"
{
#endif
/* Constructor: gives the base and size of the arena in which memory is */
/* to be carved out, note that management structures are taken from the */
/* same arena. */
void Cyg_dlmalloc_init( cyg_uint8 *base, cyg_uint32 size );

/* get some memory, return NULL if none available */
cyg_uint8 * Cyg_dlmalloc_try_alloc( cyg_uint32 size );

/* resize existing allocation, if oldsize is non-NULL, previous */
/* allocation size is placed into it. If previous size not available, */
/* it is set to 0. NB previous allocation size may have been rounded up. */
/* Occasionally the allocation can be adjusted *backwards* as well as, */
/* or instead of forwards, therefore the address of the resized */
/* allocation is returned, or NULL if no resizing was possible. */
/* Note that this differs from ::realloc() in that no attempt is */
/* made to call malloc() if resizing is not possible - that is left */
/* to higher layers. The data is copied from old to new though. */
/* The effects of alloc_ptr==NULL or newsize==0 are undefined */
cyg_uint8 * Cyg_dlmalloc_resize_alloc( cyg_uint8 *alloc_ptr, cyg_uint32 newsize,
                  cyg_uint32 *oldsize );

/* free the memory back to the pool */
/* returns true on success */
void Cyg_dlmalloc_free( cyg_uint8 *ptr );

/* Get memory pool status */
/* flags is a bitmask of requested fields to fill in. The flags are */
/* defined in common.h */
void Cyg_dlmalloc_get_status( cyg_uint32 flags,
                  struct Cyg_Mempool_Status *status );

#ifdef __cplusplus
}
#endif

#endif /* ifndef MEMPOOL_DLMALLOC_H */
/* EOF dlmalloc.h */


mpcommon.h
/*==========================================================================
//
// This file is part of the MemPool library.
// Copyright (C) 2017 Bernd Edlinger
//
// MemPool is free software; you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free
// Software Foundation; either version 3 or (at your option) any later
// version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
// for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software Foundation, Inc.,
// 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
//
// As a special exception, if other files instantiate templates or use
// macros or inline functions from this file, or you compile this file
// and link it with other works to produce a work based on this file,
// this file does not by itself cause the resulting work to be covered by
// the GNU General Public License.  However the source code for this file
// must still be made available in accordance with section (6) of the GNU
// General Public License v3.
//
// This exception does not invalidate any other reasons why a work based
// on this file might be covered by the GNU General Public License.
//
//========================================================================*/
#ifndef MEMPOOL_MPCOMMON_H
#define MEMPOOL_MPCOMMON_H

/*==========================================================================
//
//      mpcommon.h
//
//      Shared definitions used by memory allocators
//
//==========================================================================
//#####DESCRIPTIONBEGIN####
//
// Author(s):    jlarmour
// Contributors: Bernd Edlinger
// Date:         2000-06-12
// Purpose:      Shared definitions used by memory allocators
//
//
//####DESCRIPTIONEND####
//
//========================================================================*/

#if defined _WIN32 || defined _WIN32_WCE
#define MEMPOOL_WIN
#else
#define MEMPOOL_POSIX
#endif

#ifdef MEMPOOL_WIN
#define MEMPOOL_API_CALL __stdcall
#else
#define MEMPOOL_API_CALL
#endif

/* TYPE DEFINITIONS */

typedef unsigned char      cyg_uint8;
typedef unsigned int       cyg_uint32;
typedef int       cyg_bool;

/*==========================================================================
// struct Cyg_Mempool_Status is returned by the get_status() method of
// standard eCos memory allocators. After return from get_status(), any
// field of type T may be set to ((T)-1) to indicate that the information
// is not available or not applicable to this allocator.
//========================================================================*/


struct Cyg_Mempool_Status {
    const cyg_uint8 *arenabase; /* base address of entire pool */
    cyg_uint32  arenasize;      /* total size of entire pool */
    cyg_uint32  freeblocks;     /* number of chunks free for use */
    cyg_uint32  totalallocated; /* total allocated space in bytes */
    cyg_uint32  totalfree;      /* total space in bytes not in use */
    cyg_uint32  blocksize;      /* block size if fixed block */
    cyg_uint32  maxfree;        /* size of largest unused block */
    cyg_uint8   waiting;        /* are there any threads waiting for memory? */
    const cyg_uint8 *origbase;  /* address of original region used when pool */
                                /* created */
    cyg_uint32  origsize;       /* size of original region used when pool */
                                /* created */

    /*==========================================================================
    // maxoverhead is the *maximum* per-allocation overhead imposed by
    // the allocator implementation. Note: this is rarely the typical
    // overhead which often depends on the size of the allocation requested.
    // It includes overhead due to alignment constraints. For example, if
    // maxfree and maxoverhead are available for this allocator, then an
    // allocation request of (maxfree-maxoverhead) bytes must always succeed
    // Unless maxoverhead is set to -1 of course, in which case the allocator
    // does not support reporting this information.
    //========================================================================*/

    cyg_uint8   maxoverhead;
    cyg_uint32  maxAllocBytes;
    cyg_uint32  actualAllocBytes;

};

/*==========================================================================
// Flags to pass to get_status() methods to tell it which stat(s) is/are
// being requested
//========================================================================*/

#define CYG_MEMPOOL_STAT_ARENABASE       (1<<0)
#define CYG_MEMPOOL_STAT_ARENASIZE       (1<<1)
#define CYG_MEMPOOL_STAT_FREEBLOCKS      (1<<2)
#define CYG_MEMPOOL_STAT_TOTALALLOCATED  (1<<3)
#define CYG_MEMPOOL_STAT_TOTALFREE       (1<<4)
#define CYG_MEMPOOL_STAT_BLOCKSIZE       (1<<5)
#define CYG_MEMPOOL_STAT_MAXFREE         (1<<6)
#define CYG_MEMPOOL_STAT_WAITING         (1<<7)
#define CYG_MEMPOOL_STAT_ORIGBASE        (1<<9)
#define CYG_MEMPOOL_STAT_ORIGSIZE        (1<<10)
#define CYG_MEMPOOL_STAT_MAXOVERHEAD     (1<<11)

#define CYG_MEMPOOL_STAT_ALL                    \
              ( CYG_MEMPOOL_STAT_ARENABASE      \
              | CYG_MEMPOOL_STAT_ARENASIZE      \
              | CYG_MEMPOOL_STAT_FREEBLOCKS     \
              | CYG_MEMPOOL_STAT_TOTALALLOCATED \
              | CYG_MEMPOOL_STAT_TOTALFREE      \
              | CYG_MEMPOOL_STAT_BLOCKSIZE      \
              | CYG_MEMPOOL_STAT_MAXFREE        \
              | CYG_MEMPOOL_STAT_WAITING        \
              | CYG_MEMPOOL_STAT_ORIGBASE       \
              | CYG_MEMPOOL_STAT_ORIGSIZE       \
              | CYG_MEMPOOL_STAT_MAXOVERHEAD )

#endif /* ifndef MEMPOOL_MPCOMMON_H */
/* EOF mpcommon.h */
